<a href="https://colab.research.google.com/github/lephuocdat2000/-CS114.K21-/blob/master/CreateData-kelementh.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import sys
sys.setrecursionlimit(1000000)

In [2]:
import numpy as np
from numpy.random import randint
import pandas as pd
import cProfile
def partition(arr, l, r):
    x = arr[r]
    i = l
    for j in range(l, r):
        if arr[j] <= x:
            arr[i], arr[j] = arr[j], arr[i]
            i += 1
    arr[i], arr[r] = arr[r], arr[i]
    return i

# finds the kth position (of the sorted array)
# in a given unsorted array i.e this function
# can be used to find both kth largest and
# kth smallest element in the array.
# ASSUMPTION: all elements in arr[] are distinct
def kthSmallest(arr, l, r, k):
    # if k is smaller than number of
    # elements in array
    if (k > 0 and k <= r - l + 1):
        # Partition the array around last
        # element and get position of pivot
        # element in sorted array
        index = partition(arr, l, r)
        # if position is same as k
        if (index - l == k - 1):
            return arr[index]
            # If position is more, recur
        # for left subarray
        if (index - l > k - 1):
            return kthSmallest(arr, l, index - 1, k)
            # Else recur for right subarray
        return kthSmallest(arr, index + 1, r,
                           k - index + l - 1)
    return INT_MAX

#thử với k=constant

In [5]:
k=30
n=100
while(n<=7000):
  arr=np.zeros(n)
  print(n,k)
  arr=randint(0,1000000000,n)
  print(len(arr))
  print(arr)
  cProfile.run('kthSmallest(arr,0,n-1,k)')
  n+=400

1000 30
1000
[694797040 811758601 349681912 619257225 633350816  51091021 429454562
 753183567 129332739 715913136 914434144 549131034 651393913 325721204
 855884641  18390820 138600136 399552068 102347635 226136551 820660877
 658962532 450373068 560282448 202831635 136143932  74029561  12451509
 258467254 839881000 788718890 112191369 356398135 472715451 582294724
 173357638 852204352 916127332 763989521 430068069 188590484 969479746
 442799199 597764171 862633512 524348774 360188369 995702432 127142616
 362384045 602222381 554218570 233343826 773432305 180086164 624828567
 976707729 668332759 836062180 780424560 327974393  95908296 640214210
 203796360 744888528 681146456 554752678 543693100 109392926 360945062
 519115148 274985667 878976557 737861361 648013435 513563013 637425016
 872495568 350222591 742901974 579400013 426525332 401933795 212920235
 669719798 434990863 850015490 136955825 613304320 466348511  72437978
 931753093 378147607 203928696 995922023 601226838 375708530 209

In [ ]:
k=30
n=500
arr=np.zeros(n)
print(n,k)
arr=randint(0,1000000000,n)
cProfile.run('for i in range(10):kthSmallest(arr,0,n-1,k)')

500 30
         5227 function calls (2625 primitive calls) in 0.421 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
  2612/10    0.003    0.000    0.421    0.042 <ipython-input-3-bef039780bf4>:20(kthSmallest)
     2612    0.417    0.000    0.417    0.000 <ipython-input-3-bef039780bf4>:5(partition)
        1    0.000    0.000    0.421    0.421 <string>:1(<module>)
        1    0.000    0.000    0.421    0.421 {built-in method builtins.exec}
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}




26000 30
         83109 function calls (41566 primitive calls) in 245.273 seconds
100 30
         1057 function calls (540 primitive calls) in 0.034 seconds
500 30
         5227 function calls (2625 primitive calls) in 0.421 seconds


#thử với n=constant

In [ ]:
n=10000
arr=np.zeros(n)
Random_array(arr,n,1000000000)

In [ ]:
k=1
while(k<10000):
  print(k)
  runtime=cProfile.run('kthSmallest(arr,0,n-1,k)')
  k+=300

1
         16369 function calls (8187 primitive calls) in 30.093 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
     8183   30.071    0.004   30.071    0.004 <ipython-input-66-f30e93769383>:11(partition)
   8183/1    0.022    0.000   30.093   30.093 <ipython-input-66-f30e93769383>:26(kthSmallest)
        1    0.000    0.000   30.093   30.093 <string>:1(<module>)
        1    0.000    0.000   30.093   30.093 {built-in method builtins.exec}
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}


301
         16879 function calls (8442 primitive calls) in 29.022 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
     8438   29.002    0.003   29.002    0.003 <ipython-input-66-f30e93769383>:11(partition)
   8438/1    0.020    0.000   29.022   29.022 <ipython-input-66-f30e93769383>:26(kthSmallest)
        1    0.000    0.000   29

In [ ]:
  runtime=cProfile.run('kthSmallest(arr,0,n-1,10000)')

         5 function calls in 0.007 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.007    0.007    0.007    0.007 <ipython-input-66-f30e93769383>:11(partition)
        1    0.000    0.000    0.007    0.007 <ipython-input-66-f30e93769383>:26(kthSmallest)
        1    0.000    0.000    0.007    0.007 <string>:1(<module>)
        1    0.000    0.000    0.007    0.007 {built-in method builtins.exec}
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}


